src/stirling/bpf_tools/bcc_wrapper.h

```diff
@@ -480,6 +455,192 @@ class BCCWrapper {
+  }
+
+  Status SetValue(const uint32_t idx, const T& value) {
+    ebpf::StatusTuple s = underlying_->update_value(idx, value);
+    if (!s.ok()) {
+      return error::Internal(absl::Substitute(err_msg_, "set", name_, idx, s.msg()));
+    }
+    return Status::OK();
+  }
+
+ private:
+  WrappedBCCArrayTable(bpf_tools::BCCWrapper* bcc, const std::string& name) : name_(name) {
+    underlying_ = std::make_unique<U>(bcc->BPF().get_array_table<T>(name_));
+  }
+
+  const std::string name_;
+  char const* const err_msg_ = "BPF failed to $0 value for array table: $1, index: $2. $3.";
+  std::unique_ptr<U> underlying_;
+};
+
+// WrappedBCCMap includes templates parameter kUserSpaceManaged which enables the "shadow keys"
+// optimization. Files uprobe_manager.h|cc in socket tracer use this feature to speed up maps
+// that are only modified from user space.
+template <typename K, typename V, bool kUserSpaceManaged = false>
+class WrappedBCCMap {
+ public:
+  using U = ebpf::BPFHashTable<K, V>;
+
+  static std::unique_ptr<WrappedBCCMap> Create(bpf_tools::BCCWrapper* bcc,
+                                               const std::string& name) {
+    return std::unique_ptr<WrappedBCCMap>(new WrappedBCCMap(bcc, name));
+  }
+
+  size_t capacity() const { return underlying_->capacity(); }
+
+  StatusOr<V> GetValue(const K& key) const {
+    V value;
+    ebpf::StatusTuple s = underlying_->get_value(key, value);
+    if (!s.ok()) {
+      return error::Internal(absl::Substitute(err_msg_, "get", name_, s.msg()));
+    }
+    return value;
+  }
+
+  Status SetValue(const K& key, const V& value) {
+    ebpf::StatusTuple s = underlying_->update_value(key, value);
+    if (!s.ok()) {
+      return error::Internal(absl::Substitute(err_msg_, "set", name_, s.msg()));
+    }
+    if constexpr (kUserSpaceManaged) {
+      shadow_keys_.insert(key);
+    }
+    return Status::OK();
+  }
+
+  Status RemoveValue(const K& key) {
+    if constexpr (kUserSpaceManaged) {
+      if (!shadow_keys_.contains(key)) {
+        return Status::OK();
+      }
+    }
+
+    const auto s = underlying_->remove_value(key);
+    if (!s.ok()) {
+      return error::Internal(absl::Substitute(err_msg_, "remove", name_, s.msg()));
+    }
+    if constexpr (kUserSpaceManaged) {
+      shadow_keys_.erase(key);
+    }
+    return Status::OK();
+  }
+
+  std::vector<std::pair<K, V>> GetTableOffline(const bool clear_table = false) {
+    if constexpr (!kUserSpaceManaged) {
+      return underlying_->get_table_offline(clear_table);
+    }
+
+    // "r" our result.
+    std::vector<std::pair<K, V>> r;
+
+    // This is a user space managed map: we can iterate over the shadow keys.
+    for (const auto& k : shadow_keys_) {
+      auto s = GetValue(k);
+      const auto v = s.ConsumeValueOrDie();
+      r.push_back({k, v});
+      if (clear_table) {
+        PX_UNUSED(underlying_->remove_value(k));
+      }
+    }
+    if (clear_table) {
+      shadow_keys_.clear();
+    }
+    return r;
+  }
```

Comment:
Did the previous user space map implementation even provide this implementation? It appears to me this is new API surface area and didn't exist before. Assuming that understanding is correct, my preference would be to only add what is used.

Reply:
I'd like to agree with the insight here, but it is tricky. I removed the impl. and added `static_assert(kUserSpaceManaged)`. However, this static assertion failed (and I'd like to disagree with the compiler insomuch as the offending instantiation was not user space managed). Here are some other options:
1. Return a `StatusOr` from this method.
2. `ECHECK(false)` for user space managed.
3. More template magic to ensure that we don't compile this wrongly.

Reply:
ah, I see. In that case 5 is good with me.